source/common/config/grpc_mux_impl.cc

```diff
@@ -44,6 +73,10 @@ void GrpcMuxImpl::onDynamicContextUpdate(absl::string_view resource_type_url) {
 void GrpcMuxImpl::start() { grpc_stream_.establishNewStream(); }
 
 void GrpcMuxImpl::sendDiscoveryRequest(const std::string& type_url) {
+  if (shutdown_) {
```

Comment:
Do we need to hold a mutex when checking this value?

I would suggest adding a mutex in GrpcMuxImpl that protects this bool.  The AllMuxes lock should be taken before the GrpcMuxImpl mutex when setting shutdown_.

Reply:
I don't think it really changes the semantics to have a lock around the bool. If there's a race, it could still happen that just before `shutdown_ = true`, the variable is read from another thread and the message-sending code is run.

I'm not even sure this is the correct long-term solution. I'm trying to cobble something together that's good enough to make the tests pass again; I'm not planning to mark the linked issue as resolved with this PR.

Reply:
At the very least mark shutdown_ as atomic to avoid complaints from TSAN and undefined behavior on ARM which has a weaker memory model than x86.